File: MorySnip/Classes/Layers/Layer.vb
Public MustInherit Class Layer
    Public Property Offset() As Point
    Public Property Fill() As Boolean
    Public Property Pen() As Pen
    Public Property Brush() As Brush
    Public Property FirstPoint() As Point
    Public Property LastPoint() As Point

    Public Overridable Sub Start(FirstPoint As Point)
        Me.FirstPoint = FirstPoint
        Me.LastPoint = FirstPoint
    End Sub

    Public Overridable Sub [Step](StepPoint As Point)
        Me.LastPoint = StepPoint
    End Sub

    Public Overridable Sub [Stop](LastPoint As Point)
        Me.LastPoint = LastPoint
        If Me.Pen IsNot Nothing Then Me.Pen = Me.Pen.Clone()
        If Me.Brush IsNot Nothing Then Me.Brush = Me.Brush.Clone()
    End Sub

    Public Overridable ReadOnly Property Bounds() As Rectangle
        Get
            Return NormalRectingle(Me.FirstPoint, Me.LastPoint)
        End Get
    End Property

    Public Function InBounds(p As Point) As Boolean
        With Me.Bounds
            Return p.X > .X AndAlso p.Y > .Y AndAlso p.X < .X + .Width AndAlso p.Y < .Y + .Height
        End With
    End Function

    Public Overridable Sub Render(g As Graphics)
        g.CompositingQuality = Drawing2D.CompositingQuality.HighQuality
        g.InterpolationMode = Drawing2D.InterpolationMode.HighQualityBicubic
        g.SmoothingMode = Drawing2D.SmoothingMode.AntiAlias

        Me.Paint(g)
    End Sub

    Public MustOverride Sub Paint(g As Graphics)
End Class

File: MorySnip/Classes/Layers/Layer.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using Morysoft.MorySnip.Modules;

namespace Morysoft.MorySnip.Classes.Layers;

public abstract class Layer
{
    public Point Offset { get; set; }

    public bool Fill { get; set; }

    public Pen? Pen { get; set; }

    public Brush? Brush { get; set; }

    public Point FirstPoint { get; set; }

    public Point LastPoint { get; set; }

    public virtual bool IsValid => true;

    public virtual void Start(Point FirstPoint)
    {
        this.FirstPoint = FirstPoint;
        this.LastPoint = FirstPoint;
    }

    public virtual void Step(Point StepPoint) => this.LastPoint = StepPoint;

    public virtual void Stop(Point LastPoint)
    {
        this.LastPoint = LastPoint;

        if (this.Pen is not null)
        {
            this.Pen = (Pen)this.Pen.Clone();
        }

        if (this.Brush is not null)
        {
            this.Brush = (Brush)this.Brush.Clone();
        }
    }

    public virtual Rectangle Bounds => Helpers.NormalizeRectingle(this.FirstPoint, this.LastPoint);

    public bool InBounds(Point p)
    {
        return p.X > this.Bounds.X && p.Y > this.Bounds.Y && p.X < this.Bounds.X + this.Bounds.Width && p.Y < this.Bounds.Y + this.Bounds.Height;
    }

    public virtual void Render(Graphics g)
    {
        g.CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality;
        g.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
        g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.AntiAlias;

        this.Paint(g);
    }

    public abstract void Paint(Graphics g);

    public virtual void SecondaryAction() => this.Fill ^= true;

    public void SwitchPenDashStyle()
    {
        if (this.Pen is null)
        {
            return;
        }

        this.Pen.DashStyle = this.Pen.DashStyle switch
        {
            DashStyle.Solid => DashStyle.Dash,
            DashStyle.Dash => DashStyle.Dot,
            _ => DashStyle.Solid,
        };
    }
}
